fuel-asm/src/macros.rs

```diff
@@ -1,42 +1,683 @@
+//!     /// Bitwise ANDs two registers.
+//!     pub fn and(ra: u8, rb: u8, rc: u8) -> Instruction {
+//!         AND::new(check_reg_id(ra), check_reg_id(rb), check_reg_id(rc)).into()
+//!     }
+//!
+//!     // ...
+//! };
+//! ```
+//!
+//! ### Instruction Layout
+//!
+//! The function signatures of the `new` and `unpack` functions are derived from the instruction's
+//! data layout described in the `impl_instructions!` table.
+//!
+//! For example, the `unpack` method for `ADD` looks like this:
+//!
+//! ```rust,ignore
+//! // 0x10 ADD add [RegId RegId RegId]
+//! pub fn unpack(self) -> (RegId, RegId, RegId)
+//! ```
+//!
+//! While the `unpack` method for `ADDI` looks like this:
+//!
+//! ```rust,ignore
+//! // 0x50 ADDI addi [RegId RegId Imm12]
+//! pub fn unpack(self) -> (RegId, RegId, Imm12)
+//! ```
+//!
+//! ### Shorthand Constructors
+//!
+//! The shorthand instruction constructors (e.g. `add`, `and`, etc) are specifically designed to
+//! make it easier to handwrite assembly for tests or benchmarking. Unlike the `$OP::new`
+//! constructors which require typed register ID or immediate inputs, the instruction constructors
+//! allow for constructing `Instruction`s from convenient literal value inputs. E.g.
+//!
+//! ```rust
+//! use fuel_asm::{op, Instruction};
+//!
+//! // A sample program to perform ecrecover
+//! let program: Vec<Instruction> = vec![
+//!     op::move_(0x10, 0x01),     // set r[0x10] := $one
+//!     op::slli(0x20, 0x10, 5),   // set r[0x20] := `r[0x10] << 5 == 32`
+//!     op::slli(0x21, 0x10, 6),   // set r[0x21] := `r[0x10] << 6 == 64`
+//!     op::aloc(0x21),            // alloc `r[0x21] == 64` to the heap
+//!     op::addi(0x10, 0x07, 1),   // set r[0x10] := `$hp + 1` (allocated heap)
+//!     op::move_(0x11, 0x04),     // set r[0x11] := $ssp
+//!     op::add(0x12, 0x04, 0x20), // set r[0x12] := `$ssp + r[0x20]`
+//!     op::ecr(0x10, 0x11, 0x12), // recover public key in memory[r[0x10], 64]
+//!     op::ret(0x01),             // return `1`
+//! ];
+//! ```
 
-macro_rules! from_u8 {
-    ($i:ident) => {
-        impl $i {
-            /// Const fn implementation of `From<u8>`
-            pub const fn from_u8(b: u8) -> Self {
-                // Currently, the language doesn't support customized type coercion
-                //
-                // Safety: all possible values of `b` are either allocated or reserved
-                unsafe { core::mem::transmute::<u8, Self>(b) }
+/// This macro is intentionaly private. See the module-level documentation for a thorough
+/// explanation of how this macro works.
+macro_rules! impl_instructions {
+    // Recursively declares a unique struct for each opcode.
+    (decl_op_struct $doc:literal $ix:literal $Op:ident $op:ident [$($field:ident)*] $($rest:tt)*) => {
+        #[doc = $doc]
+        #[derive(Clone, Copy, Eq, Hash, PartialEq)]
+        #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+        pub struct $Op(pub (super) [u8; 3]);
```

Comment:
One of the biggest pain points with working with ops is that instead of them actually indicating which args they take it's all register so you have to constantly refer back to the docs.

I would love to go from `op::call($rA, $rB, $rC, $rD,)` to `op::call(contract_id, //etc)` .
With `OpCode` we had the chance to name the fields by turning the variants into struct variants. It seems this removes that possibility. Is there some way would could add that do this macro? It doesn't have to be within this PR but I'd like to see it be possible before approving this.

Reply:
I think we can always wrap the constructors with functions that take whatever args in whatever more human-readable layout we like - I think [these `GM` and `GTF` alternate constructors](https://github.com/FuelLabs/fuel-vm/blob/c08823e7b1ee84405779ce880bbd0d99554d26e2/fuel-asm/src/op.rs#L11-L33) might be a good example of the kind of direction we can take.

Similarly, we can also add alternatives to the `unpack` functions that "unpack" to some more human-readable type (rather than just a tuple with `RegId`s and `Imm*`s).

That's a sweet idea actually, I like the idea of adding types like that as a part of `fuel-asm` in the future like you mention (rather than doing all the register and immediate interpretation in the fuel-vm for example) as other downstream crates might find them useful / safer too.

Reply:
Yeh I really like the idea of unpacking to types. Do you think the following is possible?:
```rust
impl_instructions! {
    "Adds two registers."
    0x10 ADDI addi [Sum: RegId, Lhs: RegId, Rhs: U12]
```
Which generates something like the following:
```rust
pub fn new(sum: RegId<u64>, lhs: RegId<u64>, rhs: Imm12<U12>) -> Self

pub fn sum(&self) -> RegId<u64>
pub fn lhs(&self) -> RegId<u64>
pub fn rhs(&self) -> Imm12<U12>
```
and a debug like:
```
ADDI { sum: RegId(16), lhs: RegId(17), rhs: U12(10) }
```

Reply:
Yeah that looks sick - totally doable!